### ISSUE TITLE
[Proposal] Add a syntax to force a macro argument to be recognize as an ASTNode

### ISSUE THREAD

**Issue by Bew78LesellB** [crystal-lang/crystal#4109]
I'm having fun with macros, and making an AST pretty printer with macros, I have a specific problem that don't have any solution (or I don't know some dark language feature..):

I need to force a macro argument to be recognized as an ASTNode for analysis. 
Here is a minimal example (with a NamedArgument node) of what I need to do:
```crystal
macro ast_pp(node)
  p "{{node.class_name}}: {{node.id}}"
end

ast_pp named_arg: 42   # How to pass the NamedArgument node here ??
```
There are some other nodes that generate this issue, like `Block`, `Return`, `If`, `MultiAssign` to name the main ones.

I've found a syntax that I think is good for that, and is not used anywhere else: `%!(  )`. The macro could then be called this way:
```crystal
ast_pp %!(named_arg: 42)
```
And to help the _parser_ when trying to recognize this node, we could add an ASTNode restriction (using the Macros accessible node types), for example via a syntax like `%![Restriction](Node)`:
```crystal
ast_pp %![ArrayLiteral(ASTNode | NamedArgument)](1, "two", 3, named_arg: 42)
```

I think this syntax has 2 main advantages:
- the `%` reminds me the macros so it's clear that we can't use this syntax elsewhere.
- the `%!` characters association is not used, so the lexing/parsing should not be too hard.

What do you think ? (on the proposal, the syntax, or anything else)

**Comment by asterite** [comment]
Hi,

I'm not sure what you want here, but you can use `*args` and `**nargs` to capture all arguments to a macro, similar to how you do it with a regular method. For example:

```crystal
macro ast_pp(*args, **nargs)
  {{ p args }}
  {{ p nargs }}
end

ast_pp 1, 2, named_arg: 42 # How to pass the NamedArgument node here ??
```

**Comment by Bew78LesellB** [comment]
@asterite This is not what I want, and I agree, the simple arguments & NamedArgument are not good example..

Given this macro that generically describe an ASTNode type:
```crystal
macro ast_type(node)
  {% p "[#{node.class_name.id}]: #{node.id}" %}
end
```

### theses works:

```crystal
ast_type yield
# => [Yield]: yield

ast_type some_call 1, 2
# => [Call]: some_call(1, 2)

ast_type class A; end
# => [ClassDef]: class A\nend

ast_type 1 + 2 + 3
# => [Call]: (1 + 2) + 3
```

### theses fails:

```crystal
ast_type return true if 42 == 40 + 2
# error: void value expression

ast_type do; puts 42; end
# error: wrong number of arguments for macro 'ast_type' (given 0, expected 1)

ast_type break
# error: void value expression
```

### with new syntax:

```crystal
ast_type %!(return true if 42 == 40 + 2)
# should be: => [PostIf]: return true if 42 == (40 + 2)         (I'm not sure the type is correct will print this)

ast_type %!(break)
# should be: => [Break]: break

ast_type %!(do; puts 42; end)
# should be: => [Block]: do\n  puts(42)\nend
```
---
I need to pass any single ASTNode as first argument to a macro

Do you see the issue ? (I can give more informations on my exact use case if needed)

**Comment by asterite** [comment]
Yes, please describe your use case :-)

**Comment by bew** [comment]
Sorry for the looooong post, but I find it quite difficult to fully explain the issues, and the work I did on them..

--------------------

The original goal was to make a PEG parser generator at compile time, with nice syntactic sugar based on macros.

To learn the macro language and to see how far we can go in **user-space/macro-space** (not in **compiler-space**, so without using libraries in `"compiler/**"`), I'm making an **AST inspector** with macros only, to visually _**see**_ the AST, and the nodes I can manipulate from a macro and how.

## Simple experiment

### Simplified main macro code

Here are the main macro: (I want my code to be as generic as possible)
```crystal
# generated puts helper for a given indentation level
macro ast_puts(doc, node)
  # I do not inlude the code here as I think it is pointless

  # basically, it prints the correct indentation,
  # then based on the node.class_name, it'll either:
  # - print it directly (for simple types)
  # - print the class_name and 'recursively' call `ast_dump` with a deaper indentation level (for complex types, e.g. Call)
end

# this is the main macro:
macro ast_dump(node)

  {% if node.is_a?(Call) %} # Call

    # Call#name
    ast_puts "name", {{node.name.stringify}}

    # ... skip Call#args dump

    # Call#named_args
    {% if node.named_args %}
      {% if node.named_args.size == 1 %}
        ast_puts "named_arg", {{node.named_args.first}}
      {% else %}
        # ... skipped as not used in the usage below
      {% end %}
    {% end %}

    # Call#block
    {% if node.block %}
      ast_puts "block", {{node.block}}       # this will pass a Block node
    {% end %}

  {% elsif node.is_a?(NamedArgument) %} # NamedArgument

    # NamedArgument#name
    ast_puts "name", {{node.name}}

    # NamedArgument#value
    ast_puts "value", {{node.value}}

  {% elsif node.is_a?(Block) %} # Block

    # ... skip Block#args dump

    # Block#body
    ast_puts "body", {{node.body}}           # this will pass an Expressions ASTNode

  {% elsif node.is_a?(Expressions) %} # Expressions

    # Expressions#expressions
    {% for expr, i in node.expressions %}
      ast_puts "expr{{i + 1}}", {{expr}}     # this will pass any kind of ASTNode
    {% end %}

  {% elsif node.is_a?(If) %} # If

    # If#cond
    ast_puts "cond", {{node.cond}}

    # If#then
    ast_puts "then", {{node.then}}

  {% end %}

end
```

### usage and issues

I will use the macro with this code:
```crystal
ast_dump some_call(my_named_arg: 1 + 2) {
  return 42 if false
  puts "something"
  false
}
```

I'd like the output to look like (this is pretty close to what I already have):
```
root_node: Class:Call
| name: some_call
| named_arg: Class:NamedArgument
| | name: my_named_arg
| | value: Class:Call
| | | name: +
| | | receiver: 1
| | | arg: 2
| block: Class:Block
| | body: Class:Expressions
| | | expr1: Class:If
| | | | cond: false
| | | | then: Class:Return
| | | | | expr: 42
| | | expr2: Class:Call
| | | | name: puts
| | | | arg: "something"
| | | expr3: false
```

There are some issues with my example (some of them have workaroud):

1. `named_arg` (possible workaroud with splats) 
   The generated `ast_puts` call will be: `ast_puts "named_arg", my_named_arg: 1 + 2` 
   As you can see, the `ast_puts` does not have a `named_arg` argument, and the compiler will complain about that. 
   As you mentionned earlier, a workaround would be to use a `**nargs` splat for this specific case.

2. The Block ASTNode (possible workaroud with named block) 
  The generated `ast_puts` call will be: `ast_puts "block", do; ...some code...; end`
  Here, the compiler will complain about the unfinished arguments.
  A workaround would be to use a macro with the following signature: `macro ast_puts(doc, &block); end`

3. The `return 42` expression (no workaroud) 
  The generated `ast_puts` call from the `If` will be: `ast_puts "then", return 42`
  The compiler will just fail to parse this...
  No known workaround.

  Note: it seems that there is no macro method to get the expression of a Return ASTNode... (need another issue)

### New syntax advantages

We've seen that there is a workaround for _almost_ all cases, but I think the new syntax has a huge advantage in that 
we won't need any workaround, it'll _just work_ (and if the lexer/parser cannot understand the code, 
we can add a _type restriction_ to help)

It will allow:

1. perfect genericity when we don't know exactly the ASTNode we are manipulating (or we don't want to manage it here)
2. Easier maintainability and readability of macro receiving an ASTNode
3. Non-ambiguous way to pass ASTNodes

And maybe other things...

## My full experiment

I pushed my current AST inspector here for you to see what I have so far if you want: ![dumper.cr@c248757](https://github.com/Bew78LesellB/crast-dumper.cr/blob/c248757/src/crast/dumper.cr)

I currently haven't implemented any workaround described above (I love genericity ^^).

You may copy this file and run it (everything is self-contained ATM) 
You'll get something like this:
```
root: 42

root: nil

root: hello

root: Class:Call
| name: +
| receiver: Class:Call
| | name: +
| | receiver: 39
| | arg: 2
| arg: 1

root: class A; end   (ClassDef)

root: Class:Call
| name: new
| receiver: Type   (Path)
| arg1: 1
| arg2: 2

```

--------------

Congrat' if you read & understood all of this :-)

**Comment by ozra** [comment]
**tl;dr:** but, just in case it could be something - I copied and pasted from the file for the AST-dumper in my off-shoot project, you can cut out the pieces you want, **if this could happen to help** (the AST-dump can be reduced to like two of those funcs...) - eh, also it used "color-pollution", so cut away those color suffixes (`"some-string".yellow` etc) - or add it for colorful dumps (_much easier to parse visually_) - https://gist.github.com/ozra/3538a6715db4f76732ac175ad1352027

**Otherwise - simply ignore this.**

**Comment by bew** [comment]
I sorry for the delay, @ozra your dumper is really nice!

As I said earlier
> To learn the macro language and to see how far we can go in **user-space**/**macro-space** (not in **compiler-space**, so without using libraries in `"compiler/**"`), ....

I'd like to work only with macros (so compile-time evaluation only), without cheating with the compiler libraries :-) , to see how far we can go without going deeper in thoses libraries..

The only blocking I got is described here with this proposal, which is (I think) quite simple in the end, and would allow nice compile-time manipulation..

**Comment by ozra** [comment]
@bew I must remark that credit should go to @bcardiff - he gave me the first PoC that code has evolved from.
